js/script.js

```diff
@@ -970,3 +970,164 @@ var pushmenu = (function(){
+			const expandableCodeBlocks = [];
+			$('.content-toggle-button').on('click', function(i, button) {
+				checkCopyCodeButtonsGap();
+			})
+			$('.copy-code').each(function (index, codeBlocksItem) {
+				if (codeBlocksItem.classList) {
+					for (let i = 0; i < codeBlocksItem.classList.length; i++) {
+						if (codeBlocksItem.classList[i].startsWith("expandable-")) {
+							expandableCodeBlocks.push(codeBlocksItem);
+						}
+					}
+				}
+			});
+			expandableCodeBlocks.forEach(codeBlock => addExpandButton(codeBlock));
+			parseAllCodeBlocks();
+		})
+	});
+
+	function addExpandButton(codeBlock) {
+		const pre = $(codeBlock).find('pre').first();
+		const classes = $(codeBlock).attr('class').split(' ');
+
+		const expandableClass = classes.find(className => className.startsWith('expandable'));
+
+		if (expandableClass) {
+			let rows = parseInt(expandableClass.split('-')[1]);
+			let collapsedHeight = rows * 28 + 5;
+			pre.css('height', collapsedHeight + 'px');
+			$(codeBlock)
+			.find('pre.highlight').first().css({"margin-bottom": "45px"})
+
+			let button = $('<button class="expand-code-btn"><div class="collapsed"></div><p>expand</p></button>');
+
+			button.on('click', function () {
+				const scrolableBlock = $(button).prev().first().find('pre.highlight');
+				const clipboardButton = $(button).next('button');
+				if (button.attr('data-expanded') === 'true') {
+					pre.css('height', collapsedHeight + 'px');
+					button.attr('data-expanded', 'false');
+					button.find('p').text('expand');
+					button.get(0).scrollIntoView({ block: "center" });
+					button.find('div').removeClass('expanded');
+					button.find('div').addClass('collapsed');
+					scrolableBlock.css('overflow-y', 'auto');
+				} else {
+					if (pre.prop('scrollHeight') > 2775) {
+						$(codeBlock).find('.rouge-gutter').css("width", "60px");
+					}
+					pre.css('height', 'max-content');
+					button.attr('data-expanded', 'true');
+					button.find('p').text('collapse');
+					button.find('div').removeClass('collapsed');
+					button.find('div').addClass('expanded');
+					scrolableBlock.css('overflow-y', 'hidden');
+				}
+				clipboardButton.css('right', getCopyButtonRightGap(codeBlock));
+			});
+
+			$(codeBlock).append(button);
+		}
+	}
+
+	const clipboard = new Clipboard('.noChars');
+	clipboard.on('success', function (e) {
+		$('.clipboard-btn').removeClass('noChars');
+		e.clearSelection();
+		const trigger = e.trigger;
+		if (!$(trigger).attr('data-skip-tooltip')) {
+			showTooltip(e.trigger, 'Copied!');
+		} else {
+			$(trigger).removeAttr('data-skip-tooltip');
+		}
+	});
+
+	function getCopyButtonRightGap(scrolableBlock) {
+		const scrolableElement = $(scrolableBlock).find('pre.highlight');
+		let scrollbarWidth = scrolableElement.prop('offsetWidth') - scrolableElement.prop('clientWidth');
+		return $(scrolableBlock).height() < $(scrolableBlock).find('.rouge-code').height()
+			? scrollbarWidth >= 10 ? '12px' : '7px'
+			: '0';
+	}
```

Comment:
What is the magic function with magic numbers?

Reply:
Mozilla and Chrome have different vertical scrollbar width, so if there is Firefox then we use smaller right gap and difference between inner and outer sizes is around 12 pixels for Chrome and 0-8 pixels for Firefox (it depends from focus on scrollbar).